src/main/scala/mesosphere/marathon/core/appinfo/impl/DefaultInfoService.scala

```diff
@@ -0,0 +1,105 @@
+package mesosphere.marathon.core.appinfo.impl
+
+import mesosphere.marathon.core.appinfo._
+import mesosphere.marathon.core.appinfo.AppInfo.Embed
+import mesosphere.marathon.state._
+import org.slf4j.LoggerFactory
+
+import scala.concurrent.Future
+import scala.collection.immutable.Seq
+
+private[appinfo] class DefaultInfoService(
+    groupManager: GroupManager,
+    appRepository: AppRepository,
+    newBaseData: () => AppInfoBaseData) extends AppInfoService with GroupInfoService {
+  import scala.concurrent.ExecutionContext.Implicits.global
+
+  private[this] val log = LoggerFactory.getLogger(getClass)
+
+  override def selectApp(id: PathId, selector: AppSelector, embed: Set[AppInfo.Embed]): Future[Option[AppInfo]] = {
+    log.debug(s"queryForAppId $id")
+    appRepository.currentVersion(id).flatMap {
+      case Some(app) if selector.matches(app) => newBaseData().appInfoFuture(app, embed).map(Some(_))
+      case None                               => Future.successful(None)
+    }
+  }
+
+  override def selectBy(selector: AppSelector, embed: Set[AppInfo.Embed]): Future[Seq[AppInfo]] = {
+    log.debug(s"queryAll")
+    groupManager.rootGroup()
+      .map(_.transitiveApps.filter(selector.matches))
+      .flatMap(resolveAppInfos(_, embed))
+  }
+
+  override def selectAppsInGroup(groupId: PathId, selector: AppSelector,
+                                 embed: Set[AppInfo.Embed]): Future[Seq[AppInfo]] = {
+    log.debug(s"queryAllInGroup $groupId")
+    groupManager
+      .group(groupId)
+      .map(_.map(_.transitiveApps.filter(selector.matches)).getOrElse(Seq.empty))
+      .flatMap(resolveAppInfos(_, embed))
+  }
+
+  override def selectGroup(groupId: PathId, groupSelector: GroupSelector,
+                           appEmbed: Set[Embed], groupEmbed: Set[GroupInfo.Embed]): Future[Option[GroupInfo]] = {
+    groupManager.group(groupId).flatMap {
+      case Some(group) => queryForGroup(group, groupSelector, appEmbed, groupEmbed).map(Some(_))
```

Comment:
this method should return `None`if the returned group is not matched by the `groupSelector`.

I would recommend letting `queryForGroup` to return a `Future[Option[GroupInfo]]`as well.